fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/planning/BooleanNormalizer.java

```diff
@@ -39,46 +41,138 @@
+            } else {
+                return predicate;
+            }
         } else {
             final List<List<QueryComponent>> orOfAnd = toDNF(predicate, false);
-            return normalOr(orOfAnd.stream().map(BooleanNormalizer::normalAnd).collect(Collectors.toList()));
+            return normalOr(orOfAnd.stream().map(this::normalAnd).collect(Collectors.toList()));
         }
     }
 
-    protected static boolean needsNormalize(@Nullable final QueryComponent predicate) {
+    private boolean needsNormalize(@Nullable final QueryComponent predicate) {
         return isBooleanPredicate(predicate) &&
-            (predicate instanceof ComponentWithChildren ?
-             ((ComponentWithChildren)predicate).getChildren().stream().anyMatch(BooleanNormalizer::isBooleanPredicate) :
-             isBooleanPredicate(((ComponentWithSingleChild)predicate).getChild()));
+               (predicate instanceof ComponentWithChildren ?
+                ((ComponentWithChildren)predicate).getChildren().stream().anyMatch(this::isBooleanPredicate) :
+                isBooleanPredicate(((ComponentWithSingleChild)predicate).getChild()));
+    }
+
+    private boolean shouldNormalize(@Nullable final QueryComponent predicate) {
+        return getNormalizedSize(predicate) <= sizeLimit;
     }
 
-    protected static boolean isBooleanPredicate(@Nullable final QueryComponent predicate) {
+    private boolean isBooleanPredicate(@Nullable final QueryComponent predicate) {
         return predicate instanceof AndComponent ||
                predicate instanceof OrComponent ||
                predicate instanceof NotComponent;
     }
 
+    int getNormalizedSize(@Nullable final QueryComponent predicate) {
+        if (predicate == null) {
+            return 0;
+        }
+        return toDNFSize(predicate, false);
+    }
+
+    private int toDNFSize(@Nonnull final QueryComponent predicate, final boolean negate) {
+        if (predicate instanceof AndComponent) {
+            final List<QueryComponent> children = ((AndComponent)predicate).getChildren();
+            return negate ? orToDNFSize(children, true) : andToDNFSize(children, false);
+        } else if (predicate instanceof OrComponent) {
+            final List<QueryComponent> children = ((OrComponent)predicate).getChildren();
+            return negate ? andToDNFSize(children, true) : orToDNFSize(children, false);
+        } else if (predicate instanceof NotComponent) {
+            return toDNFSize(((NotComponent)predicate).getChild(), !negate);
+        } else {
+            return 1;
+        }
+    }
+
+    private int orToDNFSize(@Nonnull final List<QueryComponent> children, final boolean negate) {
+        return children.stream().mapToInt(p -> toDNFSize(p, negate)).sum();
+    }
+
+    private int andToDNFSize(@Nonnull final List<QueryComponent> children, final boolean negate) {
+        return children.stream().mapToInt(child -> toDNFSize(child, negate)).reduce(1, (x, y) -> x * y);
```

Comment:
Should we be worried about integer overflow?

Reply:
Hmm, good question. Probably not, but it doesn't hurt to make everything a `long` instead.

Reply:
Okay, I'm more comfortable with `long`s than `int`s, though I'm still somewhat concerned that someone will do something like give it a CNF expression with 65 children, each with 2 children, and then I think it will look like there is a size of zero. But maybe that's fine.

Reply:
Yeah, I suppose that could happen. If we're concerned about this I could add logic that checks for overflow at each step and fails the whole thing if we get there. @MMcM thoughts?

Reply:
I talked to @scgray and we concluded that this is probably worth doing right. I'll rework this to stop the computation if we hit the limit.

Reply:
Yeah, it does seem pretty simple, doesn't it. How about switching back to ints and using `Math.addExact` and `Math.multiplyExact`? I don't think they're *that* slow.

Reply:
@MMcM sure, sounds good.